runtime/marshal.js

```diff
@@ -365,13 +406,31 @@ var caml_output_val = function (){
       return this.chunk;
     }
   }
-  return function (v) {
+  return function (v, flags) {
+    flags = caml_list_to_js_array(flags);
+
+    var no_sharing = (flags.indexOf(0 /*Marshal.No_sharing*/) !== -1),
+        closures =  (flags.indexOf(1 /*Marshal.Closures*/) !== -1);
+        /* Marshal.Compat_32 is redundant since integers are 32-bit anyway */
+
+    if (closures)
+      joo_global_object.console.warn("in caml_output_val: flag Marshal.Closures is not supported.");
+
     var writer = new Writer ();
     var stack = [];
+    var intern_obj_table = new MlObjectTable();
+
+    function memo(v) {
+      var existing_offset = no_sharing ? undefined : intern_obj_table.recall(v);
+      if (existing_offset) { writer.write_shared(existing_offset); return existing_offset; }
+      else intern_obj_table.store(v);
```

Comment:
do we need to call `intern_obj_table.store` when `no_sharing` is defined ?

Reply:
We still need to update the object counter.
I can do `writer.obj_counter++` instead, and remove the assignment `writer.obj_counter = intern_obj_table.length` at the end of the function.

Reply:
obj_counter seems to only be used when sharing is enable.
https://github.com/ocaml/ocaml/blob/trunk/runtime/extern.c#L180